icefoganalytics/sfa-client
# Add Ability To Handle Zero to Many Assessments in a Given Yukon Grant Approval Letter

# Context

A given funding request letter may have zero to many assessments.
The Yukon Grant Student Approval and Yukon Grant Institution Approval need the ability to support a Funding Request that has no assessments, or possibly many assessment.

See the following files:
- `src/api/serializers/funding-requests/letters/yukon-grant-student-approval-template-serializer.ts`
- `src/api/serializers/funding-requests/letters/yukon-grant-institution-approval-template-serializer.ts`
- `src/api/templates/admin/application-letter/approval/yukon-grant-student.handlebars`
- `src/api/templates/admin/application-letter/approval/yukon-grant-institution.handlebars`
- ... others?


Also investigate possibility of zero disbursements and what that would look like in the template.
Also add some diagramming of the basic data flow of letter creation.

## Questions
- [x] How should I separate/distinguish assessment sections? Should I give assessments a title like "assessment 1" ...2 ...3? Or something else?

# Screenshots

Example of Yukon Grant Student Approval Letter without any assessments or disbursements
![image](https://github.com/icefoganalytics/sfa-client/assets/23045206/b50c78a1-dc6a-4973-a59a-e85963249f0b)

Example of letter with a two disbursements and one assessments.
![image](https://github.com/icefoganalytics/sfa-client/assets/23045206/748c2e3c-d954-4bf3-9d9a-e801e718918b)

Example of letter with two disbursements and two assessments.
![image](https://github.com/icefoganalytics/sfa-client/assets/23045206/f3fc818a-6f04-4861-89d8-19a5c052968f)

Example diagram of funding request letter template complexity
![Diagram, Funding Request Letters, 2023-08-24](https://github.com/icefoganalytics/sfa-client/assets/23045206/3c0e3ecd-46a5-409b-9029-a885fd9a3896)

# Testing Instructions

1. Boot the api and database services via `dev up`
2. Boot the web service via `cd src/web && npm run start`
3. Log in to the app via the UI at http://localhost:8080
4. Go to http://localhost:3000/api/v2/admin/funding-requests/266/letters/student.json and check that you can now serialize a letter without assessments or disbursements.
5. Go to http://localhost:3000/api/v2/admin/funding-requests/266/letters/student.html to inspect the html preview
6. Go to http://localhost:3000/api/v2/admin/funding-requests/266/letters/student.pdf to check the final pdf of the letter.
7. Go to http://localhost:3000/api/v2/admin/funding-requests/69473/letters/student.pdf for a letter with assessments.
8. Open your database editor go to the `assessment` table. Duplicated the row with `funding_request_id = 69473`
9. Go to http://localhost:3000/api/v2/admin/funding-requests/69473/letters/student.pdf and check that you now see current/previous headers around the assessment data.
